File: Theory Problems/Mostly_Sorts/BinarySearch.cpp
#include <iostream>
#include <vector>
using namespace std;


int binarySearch(const vector<int>& arr, int x){
        int left=0;
        int right=arr.size()-1;
        
        while(left<=right){
            int mid = left + (right - left) / 2; //learnt from suggestion  
            if(arr[mid]==x) return mid;
            else if(arr[mid]>x) right=mid-1;
            else left=mid+1;

            mid=(left+right)/2;
        }
        return -1;
    }
 int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector <int> arr={3,5,8,9,10,21,45};
    int x=6;
    int result=binarySearch(arr, x);
    cout<<result<<"\n";
    return 0;
 }

File: Theory Problems/Mostly_Sorts/Binary_Search.java
public class Binary_Search{
    public static void main(String [] args) {
        int[] arr = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        int target = 5;
        int result = binarySearch(arr, target);
        if (result == -1) {
            System.out.println("Element not found in the array.");
        } else {
            System.out.println("Element found at index: " + result);
        }
    }
    public static int binarySearch(int [] arr, int x){
        int left=0;
        int right=arr.length-1;
        int mid=(right+left)/2;

        while(left<=right){
            if(arr[mid]==x) return mid;
            else if (arr[mid]>x) right=mid-1;
            else left=mid+1;
            mid=(left+right)/2;
        }
        return -1;
    }
}